xarray/backends/netCDF4_.py

```diff
@@ -25,14 +26,29 @@
                   '|': 'native'}
 
 
+class VariableWrapper(object):
+    """ Ensures that variable attributes and data access operates on open datastore """
+    def __init__(self, datastore, variable_name):
+        self.var = variable_name
+        self.datastore = datastore
+
+    def __getattr__(self, attr):
+        with self.datastore.ensure_open():
+            return getattr(self.datastore.ds.variables[self.var], attr)
+
+    def __getitem__(self, key):
+        with self.datastore.ensure_open():
+            return operator.getitem(self.datastore.ds.variables[self.var], key)
+
+
 class BaseNetCDF4Array(NDArrayMixin):
     def __init__(self, variable_name, datastore):
         self.datastore = datastore
         self.variable_name = variable_name
 
     @property
     def array(self):
-        return self.datastore.ds.variables[self.variable_name]
+        return VariableWrapper(self.datastore, self.variable_name)
```

Comment:
> Are you suggesting we should basically cache array.shape and array.dtype?

Yes, exactly. We should read these in when creating these objects, when the dataset is loaded. Then we only need to go back to the underlying data on disk for pulling out array values with  `__getitem__`.

Something like:
```python
class BaseNetCDF4Array(NdimSizeLenMixin):
    def __init__(self, variable_name, datastore):
        self.datastore = datastore
        self.variable_name = variable_name
        # don't autoclose yet, because the file is already open for reading
        # variable metadata
        array = self.get_array(autoclose=False)
        self.dtype = array.dtype
        self.shape = array.shape

    def get_array(self, autoclose):
        with self.datastore.ensure_open(autoclose=autoclose):
            return self.datastore.ds.variables[self.variable_name]

    def __getitem__(self, key):
        array = self.get_array(autoclose=True)
        return array[key]

    def __array__(self, dtype=None):
        return np.asarray(self[...], dtype=dtype)
```